aten/src/ATen/core/ITensorList.h

```diff
@@ -0,0 +1,356 @@
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return *this;
+  }
+
+  ITensorListIterator operator--(int) {
+    auto old = *this;
+    TORCH_ITENSORLIST_UNWRAP(tag_, { --this_; });
+    return old;
+  }
+
+  bool operator==(const ITensorListIterator& rhs) const {
+    if (tag_ != rhs.tag_) {
+      return false;
+    }
+    TORCH_ITENSORLIST_UNWRAP(tag_, {
+      auto& rhs_it = ImplT::unwrap(rhs);
+      return this_ == rhs_it;
+    });
+  }
+
+  bool operator!=(const ITensorListIterator& rhs) const {
+    return !(*this == rhs);
+  }
+
+ private:
+  Payload payload_;
+  ITensorListTag tag_;
+};
+
+/*
+ * [Note: ITensorList]
+ * Wrapper around boxed and unboxed API containers.
+ *
+ * Tagged union of both API containers:
+ *     - `TensorList`, a.k.a. `ArrayRef<Tensor>` (the unboxed API container)
+ *     - `List<Tensor>` (the boxed API container)
+ *
+ * This container wraps around these two, without incurring in extra overhead
+ * for converting from one to another.
+ *
+ * (see https://github.com/pytorch/pytorch/issues/66328)
+ */
+class ITensorList {
+ private:
+#define DEFINE_FRIEND_CLASS(TAG, ...) friend class TORCH_ITENSORLIST_IMPL(TAG);
+  TORCH_ITENSORLIST_FORALL_TAGS(DEFINE_FRIEND_CLASS);
+#undef DEFINE_FRIEND_CLASS
+
+  using unboxed_type = TORCH_ITENSORLIST_IMPL(Unboxed)::list_type;
+  using boxed_type = TORCH_ITENSORLIST_IMPL(Boxed)::list_type;
+
+  union Payload {
+    const boxed_type* boxed;
+    unboxed_type unboxed;
+    Payload() : boxed(nullptr) {}
+    ~Payload() = default;
+  };
+
+ public:
+  using iterator = ITensorListIterator;
+  using const_iterator = ITensorListIterator;
+  using value_type = typename iterator::value_type;
+
+  ITensorList() : tag_(ITensorListTag::None) {}
+
+  ITensorList(const std::initializer_list<at::Tensor>& list)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = at::ArrayRef<at::Tensor>(list);
+  }
+
+  ITensorList(const boxed_type& boxed) : tag_(ITensorListTag::Boxed) {
+    payload_.boxed = &boxed;
+  }
+
+  ITensorList(const unboxed_type& unboxed) : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed;
+  }
+
+  template <
+      typename... UnboxedConstructorArgs,
+      typename = std::enable_if_t<
+          std::is_constructible<unboxed_type, UnboxedConstructorArgs...>::value>>
+  ITensorList(UnboxedConstructorArgs&&... args)
+      : tag_(ITensorListTag::Unboxed) {
+    payload_.unboxed = unboxed_type(std::forward<UnboxedConstructorArgs>(args)...);
+  }
+
+  size_t size() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.size(); });
+  }
+
+  bool empty() const {
+    return size() == 0;
+  }
+
+  iterator begin() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.begin(); });
+  }
+
+  iterator end() const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_.end(); });
+  }
+
+  detail::ITensorListConstRef operator[](size_t i) const {
+    TORCH_ITENSORLIST_UNWRAP(tag_, { return this_[i]; });
+  }
```

Comment:
Hot take: we shouldn't allow operator[i]. This encourages people to be naughty and repeatedly access `x[i]` multiple times when they shouldn't. How much breaks if we forbid this?

Reply:
If people really want to do this, we should probably do something like `std::vector<std::reference_wrapper<Tensor>> materialize()` which collects all the tensor references into one place so that you can index on them willy nilly, if you ABSOLUTELY ABSOLUTELY have to (but probably you just need to iterate once).

Reply:
without indexing, empty() and size() are suspect too, although I care less about them.

Reply:
> This encourages people to be naughty and repeatedly access `x[i]` multiple times when they shouldn't.

Is it because of the overhead of `switch`-ing over its tag everytime?

> If people really want to do this, we should probably do something like `std::vector<std::reference_wrapper<Tensor>> materialize()`

If that's the case, wouldn't it be better to pass it directly to the kernel (i.e. call it before dispatching to the actual kernel), instead of trusting the user to "do the right thing"?

```
Tensor wrapper_cat(const ITensorList& tensors, int64_t dim) {
  auto materialized_tensors = tensors.materialize();
  structured_cat_out_cpu_functional op;
  auto precompute = op.meta(materialized_tensors, dim);
  op.impl(materialized_tensors, precompute.dim, precompute.valid, precompute.all_contiguous, precompute.all_same_dtype,   precompute.all_same_sizes_and_stride, precompute.memory_format, *op.outputs_[0]);
  return std::move(op.outputs_[0]).take();
}
```

Reply:
> Is it because of the overhead of switch-ing over its tag everytime?

Yup. In stdlib operator[] is often unchecked. In optimizer we trust.

> If that's the case, wouldn't it be better to pass it directly to the kernel (i.e. call it before dispatching to the actual kernel), instead of trusting the user to "do the right thing"?

I don't agree, because if the user is actually just going to iterate once, doing the dynamic allocation for the vector is a pessimization.

Reply:
@ezyang any thoughts on this? Should I go ahead with this and:
- create a `materialize()` method
- kill `operator[]` and maybe `size` and `empty` (I think they would be fairly useful, though), too